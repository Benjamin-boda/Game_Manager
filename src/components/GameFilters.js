import React, { useState } from "react";
import {setTextFilter, setStatusFilter, setByConsole, sortByDate} from "../actions/filters";
import {useDispatch, useSelector} from "react-redux";

const GameFilters = () => {
    const dispatch = useDispatch()
    const gamesInStore = useSelector(state => state.games)
    const platformInStore = gamesInStore.map((game) => game.platform)
    const uniquePlatform = Array.from(new Set(platformInStore));

    const onTextChange = (e) => {
        dispatch(setTextFilter(e.target.value))
    }

    const onStatusChange = (e) => {
        dispatch(setStatusFilter(e.target.value))
    }

    const onDateChange = (e) => {
        dispatch(sortByDate(e.target.value))
    }

    const onConsoleChange = (e) => {
        dispatch(setByConsole(e.target.value))
    }

    return (
        <div className="content-container">
            <div className="input-group">
                <div className="input-group__item">
                    <input
                        className="text-input"
                        type="text" 
                        placeholder="Search games by title"
                        onChange={onTextChange}
                    />
                </div>
                <div className="input-group__item">
                    <select className="select" name="platform" onChange={onConsoleChange}>
                        <option value="">--Filter by platform--</option>
                        {
                            uniquePlatform.map((gamePlatform) => 
                                <option key={gamePlatform}>{gamePlatform}</option>
                            )
                        }
                    </select>
                </div>
                <div className="input-group__item">
                    <select className="select" name="games" onChange={onStatusChange}>
                            <option value="">--Filter by status--</option>
                            <option value="Never played">Never played</option>
                            <option value="Ongoing">Ongoing</option>
                            <option value="Finished">Finished</option>
                    </select>
                </div>
                <div className="input-group__item">
                    <select className="select" name="date" onChange={onDateChange}>
                            <option value="">--Sort by bought date--</option>
                            <option value="oldDate">Old to new</option>
                            <option value="newDate">New to old</option>
                    </select>
                </div>
            </div>
        </div>
    )
}

export default GameFilters;