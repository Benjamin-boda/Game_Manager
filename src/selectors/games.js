import moment from "moment";

//Get visible games

const getVisibleGames = (games, filters) => {
    return games.filter((game) => {
        const consoleMatch = game.platform.toLowerCase().includes(filters.console.toLowerCase());
        const textMatch = game.title.toLowerCase().includes(filters.text.toLowerCase());
        const statusMatch = game.played.toLowerCase().includes(filters.status.toLowerCase());

        return textMatch && statusMatch && consoleMatch;
    })
    .sort((a, b) => {
        if (filters.sortBy === "newDate") {
            return a.boughtAt < b.boughtAt ? 1 : -1;
        } else if (filters.sortBy === "oldDate") {
            return a.boughtAt < b.boughtAt ? -1 : 1;
        };
    })
};

export default getVisibleGames;