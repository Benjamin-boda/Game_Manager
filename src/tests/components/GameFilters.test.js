import GameFilters from "../../components/GameFilters";
import React from "react";
import { mount } from "enzyme";
import games from "../fixtures/games";
import { Provider } from "react-redux";
import * as redux from 'react-redux'
import thunk from 'redux-thunk'
import configureMockStore from "redux-mock-store";
import { shallowToJson } from "enzyme-to-json";

const middlewares = [thunk]
const mockStore = configureMockStore(middlewares);
const store = mockStore({games: games});

test("should render GameFilters page correctly", () => {
    const wrapper = shallowToJson(mount(
        <Provider store={store}>
            <GameFilters /> 
        </Provider>
        ))
    expect(wrapper).toMatchSnapshot();
});

test("should call setTextFilter", () => {
    const spy = jest.spyOn(redux, 'useDispatch')
    spy.mockReturnValue(jest.fn())
    const wrapper = mount(
        <Provider store={store}>
            <GameFilters   setTextFilter={spy}/> 
        </Provider>
        )

    wrapper.find("input").prop("onChange");

    expect(spy).toHaveBeenCalled();
});

test("should call setStatusfilter", () => {
    const spy = jest.spyOn(redux, 'useDispatch')
    spy.mockReturnValue(jest.fn())
    const wrapper = mount(
        <Provider store={store}>
            <GameFilters   setStatusFilter={spy}/> 
        </Provider>
        )

    wrapper.find("select").at(1).prop("onChange");

    expect(spy).toHaveBeenCalled();
});

test("should call setByConsole", () => {
    const spy = jest.spyOn(redux, 'useDispatch')
    spy.mockReturnValue(jest.fn())
    const wrapper = mount(
        <Provider store={store}>
            <GameFilters   setByConsole={spy}/> 
        </Provider>
        )

    wrapper.find("select").at(0).prop("onChange");

    expect(spy).toHaveBeenCalled();
});

test("should call sortByDate", () => {
    const spy = jest.spyOn(redux, 'useDispatch')
    spy.mockReturnValue(jest.fn())
    const wrapper = mount(
        <Provider store={store}>
            <GameFilters   sortByDate={spy}/> 
        </Provider>
        )

    wrapper.find("select").at(2).prop("onChange");

    expect(spy).toHaveBeenCalled();
});
