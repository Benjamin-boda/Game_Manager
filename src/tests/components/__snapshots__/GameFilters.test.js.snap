// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render GameFilters page correctly 1`] = `
<Provider
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <GameFilters>
    <div>
      <input
        onChange={[Function]}
        placeholder="Search games by title"
        type="text"
      />
      <select
        name="platform"
        onChange={[Function]}
      >
        <option
          value=""
        >
          --Filter by platform--
        </option>
        <option
          key="switch"
        >
          switch
        </option>
        <option
          key="wii"
        >
          wii
        </option>
        <option
          key="playstation-4"
        >
          playstation-4
        </option>
      </select>
      <select
        name="games"
        onChange={[Function]}
      >
        <option
          value=""
        >
          --Filter by status--
        </option>
        <option
          value="Never played"
        >
          Never played
        </option>
        <option
          value="Ongoing"
        >
          Ongoing
        </option>
        <option
          value="Finished"
        >
          Finished
        </option>
      </select>
      <select
        name="date"
        onChange={[Function]}
      >
        <option
          value=""
        >
          --Sort by bought date--
        </option>
        <option
          value="oldDate"
        >
          Old to new
        </option>
        <option
          value="newDate"
        >
          New to old
        </option>
      </select>
    </div>
  </GameFilters>
</Provider>
`;
